pkg/volume/csi/csi_metrics_test.go

```diff
@@ -0,0 +1,119 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package csi
+
+import (
+	"context"
+	"testing"
+
+	"k8s.io/kubernetes/pkg/volume"
+	//"k8s.io/kubernetes/pkg/volume/csi"
+)
+
+type VolumeStatsOptions struct {
+	VolumeSpec *volume.Spec
+
+	// this just could be volumeID
+	VolumeID string
+
+	// DeviceMountPath location where device is mounted on the node. If volume type
+	// is attachable - this would be global mount path otherwise
+	// it would be location where volume was mounted for the pod
+	DeviceMountPath string
+}
+
+func TestVolumeStats(t *testing.T) {
+	spec := volume.NewSpecFromPersistentVolume(makeTestPV("test-pv", 10, "metrics", "test-vol"), false)
+	tests := []struct {
+		name           string
+		volumeStatsSet bool
+		volumeData     VolumeStatsOptions
+		success        bool
+	}{
+
+		{
+			name:           "when nodeVolumeStats=on, VolumeID=on, DeviceMountPath=on",
+			volumeStatsSet: true,
+			volumeData: VolumeStatsOptions{
+				VolumeSpec:      spec,
+				VolumeID:        "volume1",
+				DeviceMountPath: "/foo/bar",
+			},
+			success: true,
+		},
+
+		{
+			name:           "when nodeVolumeStats=off, VolumeID=on, DeviceMountPath=on",
+			volumeStatsSet: false,
+			volumeData: VolumeStatsOptions{
+				VolumeSpec:      spec,
+				VolumeID:        "volume1",
+				DeviceMountPath: "/foo/bar",
+			},
+			success: false,
+		},
+
+		{
+			name:           "when nodeVolumeStats=on, VolumeID=off, DeviceMountPath=on",
+			volumeStatsSet: true,
+			volumeData: VolumeStatsOptions{
+				VolumeSpec:      spec,
+				VolumeID:        "",
+				DeviceMountPath: "/foo/bar",
+			},
+			success: false,
+		},
+
+		{
+			name:           "when nodeVolumeStats=on, VolumeID=on, DeviceMountPath=off",
+			volumeStatsSet: true,
+			volumeData: VolumeStatsOptions{
+				VolumeSpec:      spec,
+				VolumeID:        "volume1",
+				DeviceMountPath: "",
+			},
+			success: false,
+		},
+		{
+			name:           "when nodeVolumeStats=on, VolumeID=on, DeviceMountPath=off",
+			volumeStatsSet: true,
+			volumeData: VolumeStatsOptions{
+				VolumeSpec:      spec,
+				VolumeID:        "",
+				DeviceMountPath: "",
+			},
+			success: false,
+		},
+	}
+	for _, tc := range tests {
+
+		ctx, cancel := context.WithTimeout(context.Background(), csiTimeout)
+		defer cancel()
+
+		csiSource, _ := getCSISourceFromSpec(tc.volumeData.VolumeSpec)
+
+		csClient := setupClientWithVolumeStats(t, tc.volumeStatsSet)
+
+		_, err := csClient.NodeGetVolumeStats(ctx, csiSource.VolumeHandle, tc.volumeData.DeviceMountPath)
```

Comment:
If this function just tests `NodeGetVolumeStats`, it should be kept with other similar tests in csi_client_test.go. I was hoping that, we will have a higher level of csi metrics tests here and we won't need csi client tests then.

Reply:
Kept it in a different file, so that we can collect/capture metrics related tests ( in future) in a single place. Thanks

Reply:
I still think this test has been added to incorrect file. All tests that check `csiClient` interface belong to `csi_client_test.go` - I am not sure why this should be an exception.

This test file should instead be testing `GetMetrics` function defined in `MetricsProvider` interface. We also need actual parsing of stats info to be tested, which can be done by setting response of `NodeGetVolumeStats` in `NodeClient` object. There is already a example of it where we set `nodeGetInfoResp` in `NodeClient`.

Reply:
>I still think this test has been added to incorrect file. All tests that check csiClient interface belong to csi_client_test.go
>
Done, moved it to csi_client_test.go